Step3 리뷰 요청 드립니다.

jis1218 opened the pull request:
규남님

안녕하세요. 정인섭입니다.

Step3의 리뷰 요청을 드립니다.

감사합니다!

jis1218 committed:
doc(README.md): Step1 요구사항 작성

- Step1 요구사항 작성

jis1218 committed:
feat(Entity): Step1 요구사항에 맞게 Entity에 annotation 추가

- Entity에 annotation 추가
- BaseEntity로 createdAt, updatedAt 관리

jis1218 committed:
test(AnswerTest): Answer Entity 테스트 코드 작성

- Answer Entity 테스트 코드 작성

jis1218 committed:
test(AnswerTest): 테스트 코드 추가

- Answer관련 테스트 코드 추가
- UserTest에 데이터 하나 더 추가

jis1218 committed:
docs(application.properties) - sql문 확인 기능, ddl-auto 기능 추가

jis1218 committed:
test(AnswerTest): AnswerTest에 검증 코드 추가

findById로 찾은 객체와 id를 제공한 객체가 동일한지 확인

jis1218 committed:
test(QuestionTest): QuestionTest 테스트 케이스 추가

- QuestionTest 테스트 케이스 추가
- Question 클래스에 기본 생성자 추가

jis1218 committed:
test(UserTest): UserTest 테스트 케이스 추가

- UserTest 테스트 케이스 추가
- User 클래스 잘못된 annotation 수정

jis1218 committed:
refactor: Entity 클래스에 @NonNull 삭제

- Entity 클래스에 @NonNull 삭제
- @Column에 nullalbe=false 적용

jis1218 committed:
doc(application.properties): JPA 쿼리 파라미터 확인 설정 추가

jis1218 committed:
test: @DirtiesContext 어노테이션 추가 및 findById 테스트 제거

jis1218 committed:
refactor: User 클래스에 id가 들어가는 생성자 삭제

jis1218 committed:
refactor: BaseEntity에 createdAt의 nullable = false 삭제

jis1218 committed:
docs(application.properties): newline 추가

jis1218 committed:
fix(bug fix): createdAt에 값 안들어가는 문제 해결, nullable = false 복구

jis1218 committed:
refactor: 필요없는 import 제거

jis1218 committed:
refactor: 연관관계 재정의에 따른 수정

jis1218 committed:
Merge branch 'step2'

# Conflicts:
#	src/main/java/qna/domain/Answer.java
#	src/main/java/qna/domain/Question.java
#	src/test/java/qna/domain/AnswerTest.java

jis1218 committed:
docs(README.md): Step2 기능목록 추가

jis1218 committed:
refactor: 각 엔터티 연관관계 재설정에 따른 수정

엔터티 연관관계 재설정에 따라 테스트 및 클래스 수정

jis1218 committed:
refactor: Fetch 방법 변경 및 @Where 추가에 따른 메서드명 변경

- @ManyToOne의 Fetch 방법 변경(Answer, Question 클래스)
- @Where 추가에 따른 메서드명 변경

jis1218 committed:
refactor: Answer 클래스에 @Where 추가에 따른 리팩터링

jis1218 committed:
docs(README.md): Step3 기능목록 추가

jis1218 committed:
refactor: Answer 클래스에 답변 작성자 확인하는 기능 추출

- Answer 클래스에 답변 작성자 확인하는 기능 추출
- 기능 추출에 따른 테스트 클래스 추가

jis1218 committed:
refactor: Question 클래스에 질문 작성자 확인하는 기능 추출

- Question 클래스에 답변 작성자 확인하는 기능 추출
- 기능 추출에 따른 테스트 클래스 추가

jis1218 committed:
refactor: 테스트 클래스 이름변경

- 테스트 클래스의 성격을 명확히 하기 위해 클래스명 변경

jis1218 committed:
refactor: findByIdAndDeletedFalse 다시 복구

jis1218 committed:
refactor: equals 메서드 수정(날짜 비교 삭제)

jis1218 committed:
refactor: equals, hashcode 추가

jis1218 committed:
refactor: QnAService 및 테스트 코드 리팩터링

jis1218 committed:
refactor: 일급 컬렉션 추가하여 QnAService 리팩터링

jis1218 committed:
Merge Conflicts

jis1218 committed:
style(formatting) : formatting

jis1218 committed:
test(AnswersTest): Answers 클래스 생성에 따른 AnswersTest 추가

- 답변 중 다른 사람이 쓴 글이 있을 때 validation 진행

kyucumber commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public List<DeleteHistory> deleteHistories() {
+        return deleteHistories;
```
외부에 이렇게 노출되면 add 등을 수행할 수 있지 않을까요?

`Collections.unmodifibleList` 등을 활용해보시면 어떨까요?

다른 클래스들도 동일하게 되어있는데 한번 확인해주시면 좋을 것 같습니다.

jis1218 replied:
네 수정하였습니다. 외부에 노출되는 것까지 신경쓰면서 자료구조를 선택해야 한다는 것을 배웠습니다. 감사합니다!

kyucumber commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -29,34 +29,24 @@ public QnAService(QuestionRepository questionRepository, AnswerRepository answer
+
+        question.validateQuestionOwner(loginUser, question);
+
+        Answers answers = new Answers(answerRepository.findByQuestion(question));
+
+        answers.validateOwner(loginUser);
+
+        DeleteHistories deleteHistories = new DeleteHistories();
+        deleteHistories.addDeletedQuestion(question);
+        deleteHistories.addDeletedAnswers(answers);
```
`Answer`, `Question` 내부에서 `DeleteHistory`를 생성하도록 구성하지 않으셨네요.

만약 위에 작성한 케이스와 같이 관계를 맺는 경우 `DeleteHistory`와 `Answer`, `Question` 간에 결합이 생기게 됩니다.

`Answer` -> `DeleteHistory`
`Question` -> `DeleteHistory`

이 부분의 설계는 개인차가 있겠지만 `DeleteHistory`의 변경에 여러 엔티티가 영향을 받을 수 있어 서비스 레이어의 코드가 조금 더 많아지더라도 위에 구현해주신 것 같이 하는게 더 설계상 이점이 있다고 생각됩니다. 👍 

만약에 이 부분에 대해서도 분리하고 싶다면 `EventListener` 등을 활용할 수도 있겠네요.

jis1218 replied:
수정의 영향을 적게 받는 설계의 중요성에 대해 다시 한번 느낍니다!

kyucumber commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DataJpaTest
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+public class AnswerRepositoryTest {
+    public static Answer A1;
+    public static Answer A2;
+    public static Answer A3;
+
+    @Autowired
+    private AnswerRepository answerRepository;
+
+    @Autowired
+    private QuestionRepository questionRepository;
+
+    @Autowired
+    private UserRepository userRepository;
+
+    private Answer a1;
+    private Answer a2;
+    private Answer a3;
+
+    private Question q1;
+
+    @BeforeEach
+    void setUp() {
+
+        User javajigi = userRepository.save(UserRepositoryTest.JAVAJIGI);
+        User sanjigi = userRepository.save(UserRepositoryTest.SANJIGI);
+        User insup = userRepository.save(UserRepositoryTest.INSUP);
+
```
```
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)<br class="Apple-interchange-newline">
```

위 어노테이션을 붙여주신 이유가 있을까요?

코드를 살펴보았을 때 다른 클래스에 있는 static 변수를 참조하면서 테스트 컨텍스트 간 auto increment 아이디의 충돌 등이 발생할 것으로 보이네요.

이를 해결하기 위해 위 어노테이션을 붙이신게 맞을까요?

각각의 테스트는 여러 컨텍스트간 영향을 주지 않도록 설계를 해주셔야 합니다.

static 변수를 여러 클래스에서 참조하는 부분은 지양해주시는게 좋을 것 같습니다.

테스트 간 영향을 주지 않도록 변경해주시고, 불필요한 어노테이션도 제거해주시면 좋을 것 같습니다.

jis1218 replied:
전에 annotation을 안붙였을 때 에러가 났던 이유가 static 변수를 사용했기 때문이라는 사실을 알았네요. static 변수를 없애고 각 테스트 별로 변수를 가져가도록 하였습니다. 덕분에 DirtiesContext 어노테이션을 삭제하였는데 id값은 올라가지만 테스트에 영향을 주진 않는 것 같네요~

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

jis1218 committed:
refactor : List 구현체 변경, Question, Answer에게 DeleteHistory 객체 생성 기능 위임

- 일급 컬렉션(Answers, DeleteHistories)의 List return을 unmodifiableList로 변경
- Question, Answer에 DeleteHistory 객체 생성 기능 위임, DeleteHistories와 의존관계 삭제

jis1218 committed:
refactor : Answer, Question의 삭제 메서드 이름 변경에 따른 테스트 코드 변경

jis1218 committed:
refactor: 테스트 코드 내의 static 변수 삭제, @DirtiesContext 삭제

- 테스트 코드 내의 static 변수 삭제와 그의 따른 refactoring
- 각 RepositoryTest에 @DirtiesContext 삭제

kyucumber reviewed (approved):
마지막 단계까지 진행하시느라 고생 많으셨어요.

다음 미션 진행을 위해서 머지할게요~
